#include "ManageInventory.h"
#include "ui_ManageInventory.h"
#include "EditProductDialog.h"
#include "Food.h"
#include "Beverage.h"
#include "HouseholdItem.h"
#include "Exceptions.h"
#include <QMessageBox>
#include <QHeaderView>

ManageInventory::ManageInventory(Store* store, QWidget *parent)
    : QDialog(parent), ui(new Ui::ManageInventory), m_store(store)
{
    ui->setupUi(this);
    setupTable();
    loadProducts();

    connect(ui->tableProducts, &QTableView::doubleClicked, this, &ManageInventory::onProductDoubleClicked);
    connect(ui->btnDelete, &QPushButton::clicked, this, &ManageInventory::onDeleteProductClicked);
    connect(ui->btnRefresh, &QPushButton::clicked, this, &ManageInventory::onRefreshClicked);
    connect(ui->txtSearch, &QLineEdit::textChanged, this, &ManageInventory::onSearchTextChanged);
    connect(ui->cmbFilter, QOverload<int>::of(&QComboBox::currentIndexChanged), this, &ManageInventory::onFilterChanged);
}

ManageInventory::~ManageInventory()
{
    delete ui;
}

void ManageInventory::setupTable()
{
    m_model = new QStandardItemModel(this);
    m_model->setColumnCount(8);
    m_model->setHeaderData(0, Qt::Horizontal, "M√£ SP");
    m_model->setHeaderData(1, Qt::Horizontal, "T√™n S·∫£n Ph·∫©m");
    m_model->setHeaderData(2, Qt::Horizontal, "Lo·∫°i");
    m_model->setHeaderData(3, Qt::Horizontal, "Gi√° (ƒë)");
    m_model->setHeaderData(4, Qt::Horizontal, "S·ªë L∆∞·ª£ng");
    m_model->setHeaderData(5, Qt::Horizontal, "Th·ªÉ T√≠ch (ml)");
    m_model->setHeaderData(6, Qt::Horizontal, "H·∫°n SD");
    m_model->setHeaderData(7, Qt::Horizontal, "B·∫£o H√†nh (th√°ng)");

    ui->tableProducts->setModel(m_model);
    ui->tableProducts->setEditTriggers(QAbstractItemView::NoEditTriggers);
    ui->tableProducts->setSelectionBehavior(QAbstractItemView::SelectRows);
    ui->tableProducts->setSelectionMode(QAbstractItemView::SingleSelection);
    ui->tableProducts->horizontalHeader()->setSectionResizeMode(QHeaderView::Stretch);
    ui->tableProducts->setAlternatingRowColors(true);
}

void ManageInventory::loadProducts()
{
    loadProductsFiltered(0, "");
}

void ManageInventory::loadProductsFiltered(int filterType, const QString& searchText)
{
    m_model->removeRows(0, m_model->rowCount());

    QString search = searchText.trimmed().toLower();

    m_store->forEachProduct([&](const QString&, Product* p) {
        if (!p) return;

        Food* f = dynamic_cast<Food*>(p);
        Beverage* b = dynamic_cast<Beverage*>(p);
        HouseholdItem* h = dynamic_cast<HouseholdItem*>(p);

        // L·ªçc theo lo·∫°i
        bool typeMatch = false;
        if (filterType == 0) typeMatch = true;
        else if (filterType == 1 && f) typeMatch = true;
        else if (filterType == 2 && b) typeMatch = true;
        else if (filterType == 3 && h) typeMatch = true;

        if (!typeMatch) return;

        // L·ªçc theo t√¨m ki·∫øm
        if (!search.isEmpty()) {
            QString name = p->getName().toLower();
            QString id = p->getId().toLower();
            if (!name.contains(search) && !id.contains(search))
                return;
        }

        QList<QStandardItem*> row;
        row << new QStandardItem(p->getId());
        row << new QStandardItem(p->getName());

        QString typeName;
        if (f) typeName = "ƒê·ªì ƒÉn";
        else if (b) typeName = "Th·ª©c u·ªëng";
        else if (h) typeName = "ƒê·ªì gia d·ª•ng";
        row << new QStandardItem(typeName);

        row << new QStandardItem(QString::number(p->getBasePrice(), 'f', 0));
        row << new QStandardItem(QString::number(p->getQuantity()));
        row << new QStandardItem(b ? QString::number(b->getVolume(), 'f', 0) : "");

        QString expiryDate = "";
        if (f) expiryDate = f->getExpiryDate();
        else if (b) expiryDate = b->getExpiryDate();
        row << new QStandardItem(expiryDate);

        row << new QStandardItem(h ? QString::number(h->getWarrantyMonths()) : "");

        m_model->appendRow(row);
    });

    ui->lblTotal->setText(QString("T·ªïng s·ªë s·∫£n ph·∫©m: %1").arg(m_model->rowCount()));
}

Product* ManageInventory::getProductFromRow(int row)
{
    if (row < 0 || row >= m_model->rowCount())
        return nullptr;

    QString productId = m_model->item(row, 0)->text();
    return m_store->findProductById(productId);
}

void ManageInventory::onProductDoubleClicked(const QModelIndex &index)
{
    if (!index.isValid()) return;

    Product* p = getProductFromRow(index.row());
    if (!p) return;

    EditProductDialog dialog(p, this);

    if (dialog.exec() == QDialog::Accepted) {
        try {
            // C·∫≠p nh·∫≠t th√¥ng tin s·∫£n ph·∫©m
            QString newName = dialog.getName();
            double newPrice = dialog.getPrice();
            int newQuantity = dialog.getQuantity();

            if (!newName.isEmpty()) {
                p->setName(newName);
            }

            // üõ°Ô∏è BACKEND VALIDATION: Ch·∫∑n s·ªë √¢m (kh√¥ng tin UI)
            if (newQuantity < 0) {
                QMessageBox::critical(this, "L·ªói", 
                    "S·ªë l∆∞·ª£ng kh√¥ng th·ªÉ √¢m! (C√≥ th·ªÉ UI b·ªã bypass)");
                return;
            }

            p->setBasePrice(newPrice);
            p->setQuantity(newQuantity);

            // C·∫≠p nh·∫≠t c√°c tr∆∞·ªùng ƒë·∫∑c bi·ªát theo lo·∫°i s·∫£n ph·∫©m
            if (Food* f = dynamic_cast<Food*>(p)) {
                QString newExpiry = dialog.getExpiryDate();
                try {
                    Validation::validateDateDDMMYYYY(newExpiry);
                    f->setExpiryDate(newExpiry);
                } catch (const InvalidDateException& e) {
                    QMessageBox::warning(this, "C·∫£nh b√°o",
                                         QString("Ng√†y kh√¥ng h·ª£p l·ªá: %1").arg(e.what()));
                }
            }
            else if (Beverage* b = dynamic_cast<Beverage*>(p)) {
                QString newExpiry = dialog.getExpiryDate();
                try {
                    Validation::validateDateDDMMYYYY(newExpiry);
                    b->setExpiryDate(newExpiry);
                } catch (const InvalidDateException& e) {
                    QMessageBox::warning(this, "C·∫£nh b√°o",
                                         QString("Ng√†y kh√¥ng h·ª£p l·ªá: %1").arg(e.what()));
                }
                b->setVolume(dialog.getVolume());
            }
            else if (HouseholdItem* h = dynamic_cast<HouseholdItem*>(p)) {
                h->setWarrantyMonths(dialog.getWarranty());
            }

            QMessageBox::information(this, "Th√†nh c√¥ng",
                                     "ƒê√£ c·∫≠p nh·∫≠t th√¥ng tin s·∫£n ph·∫©m th√†nh c√¥ng!");

            loadProductsFiltered(ui->cmbFilter->currentIndex(), ui->txtSearch->text());

        } catch (const std::exception& e) {
            QMessageBox::critical(this, "L·ªói",
                                  QString("Kh√¥ng th·ªÉ c·∫≠p nh·∫≠t: %1").arg(e.what()));
        }
    }
}

void ManageInventory::onDeleteProductClicked()
{
    QModelIndex index = ui->tableProducts->currentIndex();
    if (!index.isValid()) {
        QMessageBox::warning(this, "C·∫£nh b√°o", "Vui l√≤ng ch·ªçn s·∫£n ph·∫©m c·∫ßn x√≥a!");
        return;
    }

    Product* p = getProductFromRow(index.row());
    if (!p) return;

    QMessageBox::StandardButton reply;
    reply = QMessageBox::question(this, "X√°c nh·∫≠n x√≥a",
                                  QString("B·∫°n c√≥ ch·∫Øc mu·ªën x√≥a s·∫£n ph·∫©m '%1'?\n\n"
                                          "‚ö†Ô∏è C·∫£nh b√°o: H√†nh ƒë·ªông n√†y kh√¥ng th·ªÉ ho√†n t√°c!\n"
                                          "S·∫£n ph·∫©m s·∫Ω b·ªã x√≥a ho√†n to√†n kh·ªèi h·ªá th·ªëng.")
                                      .arg(p->getName()),
                                  QMessageBox::Yes | QMessageBox::No);

    if (reply == QMessageBox::Yes) {
        QString productId = p->getId();
        QString productName = p->getName();

        // ‚úÖ TODO: Implement Store::deleteProduct method
        // For now, show warning that this needs Store-level implementation
        QMessageBox::warning(this, "Ch·ª©c nƒÉng ch∆∞a ho√†n thi·ªán",
            QString("‚ö†Ô∏è X√ìA S·∫¢N PH·∫®M C·∫¶N TRI·ªÇN KHAI ·ªû STORE LEVEL\n\n"
                    "C·∫ßn implement:\n"
                    "1. Store::deleteProduct(productId)\n"
                    "2. X√≥a kh·ªèi HashTable (productById, productByName)\n"
                    "3. delete p; // Gi·∫£i ph√≥ng b·ªô nh·ªõ\n"
                    "4. Product::unregisterUsedId(productId)\n\n"
                    "‚ùå HI·ªÜN T·∫†I: Ch·ª©c nƒÉng b·ªã v√¥ hi·ªáu h√≥a ƒë·ªÉ tr√°nh memory leak!"));

        loadProductsFiltered(ui->cmbFilter->currentIndex(), ui->txtSearch->text());
    }
}

void ManageInventory::onRefreshClicked()
{
    loadProductsFiltered(ui->cmbFilter->currentIndex(), ui->txtSearch->text());
    QMessageBox::information(this, "Th√¥ng b√°o", "ƒê√£ l√†m m·ªõi d·ªØ li·ªáu!");
}

void ManageInventory::onSearchTextChanged(const QString &text)
{
    loadProductsFiltered(ui->cmbFilter->currentIndex(), text);
}

void ManageInventory::onFilterChanged(int index)
{
    loadProductsFiltered(index, ui->txtSearch->text());
}
